DS-3447 : ORCID v2 integration (using DSpace/orcid-jaxb-api)

tdonohue opened the pull request:
This PR is an alternate to #2028, including only the changes needed to core DSpace classes

It cherry-picks the cored DSpace API changes from #2028, *excluding* any of the added `org.orcid.*` classes in favor of https://github.com/DSpace/orcid-jaxb-api

It adds in a new dependency on `orcid-jaxb-api` version `2.1.0-SNAPSHOT`. This `2.1.0-SNAPSHOT` dependency is the exact code generated by https://github.com/DSpace/orcid-jaxb-api/pull/1

Additional minor code refactoring was necessary to align code with the `orcid-jaxb-api`, as a few classes/methods have different names now.

**This PR has not yet been tested, but it does compile.**

DO NOT MERGE UNTIL THE FOLLOWING IS COMPLETE:

- [x] This PR must be tested fully to ensure it provides the same behavior as #2028.
- [x] Assuming tests succeed, then https://github.com/DSpace/orcid-jaxb-api/pull/1 must be merged
- [x] An official `2.1.0` release of `orcid-jaxb-api` should be cut & sent to Maven Central
- [x] This PR's POM dependency should be updated to use that official release.
- [x] Finally, this PR can be merged, once all the above is completed.

Pinging @jonas-atmire , @benbosman and anyone else interested for reviews / testing / feedback.

jonas-atmire committed:
DS-3447 Orcid v2 implementation + removal of obsolete v1 implementation

jonas-atmire committed:
DS-3447 Additional licences + Default orcid.api.url

jonas-atmire committed:
DS-3447 Removal of more unused ORCID v1 classes + commenting in new classes

jonas-atmire committed:
DS-3447 Removal of incorrectly copied <p> tag + re-add accidentally removed license header

tdonohue committed:
Update/Refactor to use DSpace/orcid-jaxb-api (currently SNAPSHOT version)

kshepherd commented:
Travis-CI build failed, but it was not a real error, I don't think (the case of the mysterious disappearing dspace API dependency) so I've manually restarted the build to see if that helps..

kshepherd commented on dspace-api/src/main/java/org/orcid/utils/DateUtils.java:
```diff
@@ -0,0 +1,177 @@
+ * the software.
+ *
+ * =============================================================================
+ */
```
This license is triggering an error in our license-check task... did we expect this utils package to be included with the jaxb packaged jar, or is this a set of supplementary utils?

[ERROR] Failed to execute goal com.mycila:license-maven-plugin:2.11:check (check-headers) on project dspace-api: Some files do not have the expected license header -> [Help 1]

tdonohue replied:
Hmmm...looks like a few `org.orcid.utils.*` files accidentally snuck in here. Thanks for pointing these both out, @kshepherd! I don't believe either is need in our source code. But, I'll look into it and get it resolved.

kshepherd commented on dspace-api/src/main/java/org/orcid/utils/ReleaseNameUtils.java:
```diff
@@ -0,0 +1,38 @@
+package org.orcid.utils;
```
This file needs a valid DSpace license header

[ERROR] Failed to execute goal com.mycila:license-maven-plugin:2.11:check (check-headers) on project dspace-api: Some files do not have the expected license header -> [Help 1]

tdonohue replied:
Another `org.orcid.utils.*` file that shouldn't be here.  I'll get it removed.

kshepherd reviewed (commented):
Just noting license issues (the root cause of Travis CI problems). I'll continue testing on my local VM in the meantime.

tdonohue committed:
Remove unnecessary org.orcid.* classes (accidentally included in a previous commit)

tdonohue commented:
@kshepherd : thanks for the notes you made here. This PR has now been fixed, as I removed the unnecessary `org.orcid.utils.*` classes that were accidentally included.

kshepherd commented on dspace-api/src/main/java/org/dspace/authority/orcid/Orcidv2.java:
```diff
@@ -0,0 +1,145 @@
+
+    public void init() throws IOException {
+        if (StringUtils.isNotBlank(accessToken) && StringUtils.isNotBlank(clientSecret)) {
+            String authenticationParameters = "?client_id=" + clientId + "&client_secret=" + clientSecret + "&scope=/read-public&grant_type=client_credentials";
```
Where do these credentials (clientId, clientSecret) come from? Are there some config defaults I'm missing? I can't see reference to them in the spring/api/orcid-authority-services.xml spring config or any other config files
(@jonas-atmire , can you shed some light here?)

kshepherd commented:
When searching in the AJAX popup, whether from Submission Form or an Item Edit screen, I get an unmarshall error. I've tested to make sure I can reach the ORCID v2 API endpoint and that seems fine... 
Looking at [https://members.orcid.org/api/tutorial/search-orcid-registry] (v2.1 examples), I've realised that though clientId and clientSecret are referred to and passed in the commits included here, there's no config reference to them. I assume they should be coming in via spring? Am I missing something?

`2018-05-02 02:43:35,424 ERROR org.dspace.content.authority.SolrAuthority @ Error
java.lang.RuntimeException: Unable to unmarshall orcid messagejavax.xml.bind.UnmarshalException: unexpected element (uri:"", local:"html"). Expected elements are <
{http://www.orcid.org/ns/search}search>
        at org.dspace.authority.orcid.xml.Converter.unmarshall(Converter.java:42)
        at org.dspace.authority.orcid.xml.XMLtoBio.convert(XMLtoBio.java:44)
        at org.dspace.authority.orcid.Orcidv2.queryBio(Orcidv2.java:163)
        at org.dspace.authority.orcid.Orcidv2.queryBio(Orcidv2.java:174)
        at org.dspace.authority.orcid.Orcidv2.queryAuthorities(Orcidv2.java:107)
        at org.dspace.content.authority.SolrAuthority.addExternalResults(SolrAuthority.java:149)
        at org.dspace.content.authority.SolrAuthority.getMatches(SolrAuthority.java:120)
        at org.dspace.content.authority.SolrAuthority.getMatches(SolrAuthority.java:185)
        at org.dspace.content.authority.ChoiceAuthorityServiceImpl.getMatches(ChoiceAuthorityServiceImpl.java:124)
        at org.dspace.app.xmlui.cocoon.AJAXMenuGenerator.generate(AJAXMenuGenerator.java:84)`

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -1966,7 +1966,8 @@ mail.helpdesk = ${mail.admin}
 
-
+# URL of ORCID API
+orcid.api.url = https://api.orcid.org/v2.0
```
I believe this is the source of the error that @kshepherd noted.  This should be set to https://pub.orcid.org/v2.0 by default, as that's the *public* API.  The api.orcid.org path is for Members only, and requires authentication.

tdonohue committed:
Fix default ORCID API URL. Move to correct place in dspace.cfg

tdonohue commented:
@kshepherd : I've fixed the `UnmarshalException` in the latest commit.  The ORCID API URL was pointing at the private (members-only) API. I switched it to the public one here: https://github.com/DSpace/DSpace/pull/2039/commits/8c1580bbe51dd1b7fbcefecf29536d092362052d

That said, even with this change, I'm not getting back ORCID search results (but I don't see any errors in the logs).  Still trying to debug why this isn't working.

I also agree with you -- I'm confused about whether we need a `clientID` or `clientSecret` to authenticate with the ORCID API prior to searching, etc.  So, I'm hoping @jonas-atmire can find time to help us figure out if this is required or not.

kshepherd commented:
@tdonohue the ORCID 2.1 API docs suggest client ID and secret are needed for the public API as well (just a regular free API account). If we don't hear back from atmire devs, when I have time I'll try and reverse engineer the bean to figure out what the spring config should be

tdonohue commented:
@kshepherd : yes, I saw that in the ORCID 2.1 API docs too. But, using the API via GET (in your browser) works perfectly fine without authentication.  Try it out, the API is completely browseable/usable via Swagger: https://pub.orcid.org/v2.1/#/Public_API_v2.1  I've tried the `/search?q=[query]` endpoint myself and it seems to return real results without authentication.  

So, I guess I'm not yet convinced that we have an authentication problem -- it's possible there's an issue in the code itself here.  Still trying to find time myself to debug.

tdonohue committed:
Bug fixes to querying for ORCIDs by name

tdonohue commented:
@kshepherd : I've found a few bugs in the code that affected querying for ORCIDs by a person's name.  It seems the behavior is slightly different in `dspace/orcid-jaxb-api` than it was in the original PR.

The fixes are in this new commit: b8744f34657b5eaaac4fbc7016ec6d9780c61ea2

As of now, I'm able to query for ORCIDs by name using the "Lookup & Add" popup in the XMLUI submission process.  So, that's working now!  I haven't fully tested all ORCID features, but that's a step in the right direction

kshepherd commented:
@tdonohue : Great, I too can see results in the submissions and edit item form when searching by name, and when searching by ORCID ID *if* the ID is encapsulated in quotes.

I have a few suggestions for improving performance/usability, but if they're out of scope for this simple "make it work again" patch then that's fine...
- Only fetch bios for results on *this page*... most searches that include a common name will return the full 100 (default max) rows and it looks to me like we're fetching bios for all rows as we send the full query, not just the 10 that we're displaying on the current page? How much work is it to change the query bio behaviour so it's called for result set to be displayed, not the full result set returned from ORCID? 
- When fetching bios, it looks like we attempt to get optional info like email and country, etc., which is quite helpful for identifying people when duplicate names could be returned, but we don't ever display it in the (XMLUI) preview of search results when looking up names/IDs. If we included this in the search results I think it'd save repository users/staff a lot of time and avoid errors in selecting the wrong ORCID record
- The last change to allow query by name looks good, but it has broken 'search by ID' a little bit, which we certainly should discourage as it's a far better/quicker way of getting a successful result if we do happen to know the author's ORCID ID. I can make it work by encapsulating the ID in quotes, but I think maybe we could help by detecting some common patterns and if it looks like an ORCID ID instead of a name, encapsulating it in quotes or performing a direct 'by id' search?

kshepherd commented:
@tdonohue on further reflection, of the comments I listed above, only the "help search by ID to work better/quicker" is really relevant for this PR, so I guess we can just consider the other comments as a start for a subsequent "improve ORCID authority" PR after this is merged. If you like, I can commit a small update to this PR to look for patterns that match ORCID IDs / URLs and encapsulate them in quotes to get them working better, otherwise I'll just approve this as is!

tdonohue commented:
Hi @kshepherd : I think your ideas for improvements sound reasonable...however, I'm going to admit that my goal in this PR was to simply get #2028 to work with https://github.com/DSpace/orcid-jaxb-api   

95+% of this code was written by Atmire, so I'm not sure I can or should speak for it. I can only speak for the minor bug fixes I applied and the configuration that I changed to get it hooked up to `orcid-jaxb-api`. (As a sidenote: the "query by name" fix was applied because, without it, query by name wasn't working at all. It's possible there's a better way to fix that code, but I just went for a quick fix to get it working)

So, if you have time to work more on improving this PR (you should have edit access), I'd be glad to give that work a code review/test.  But, I'm not sure I'll have much time in the next few weeks (the DSpace 7 Sprint has started) to do more work here myself.  I'd just encourage we try and scope this PR to "getting ORCID working again" (as that's the real v6.3 blocker here).  More enhancements can come in future PRs.

kshepherd reviewed (approved):
+1, my testing is successful, I can search for a person by name (or by ORCID in quotes) and they are successfully added to the authority index, so from my perspective this PR cleanly takes us back to where we were when the ORCID v1 API was working with DSpace authority.

I'll address my other suggestions in separate, subsequent PRs

tdonohue committed:
Update POM with newly released version of orcid-jaxb-api

tdonohue commented:
@kshepherd : I just updated the POM in this PR to point at the newly released [`orcid-jaxb-api` version 2.1.0](https://github.com/DSpace/orcid-jaxb-api/releases/tag/orcid-jaxb-api-2.1.0)

I reran my tests on my end, and it looks to still all work great. So, I think this PR is ready to be merged.

kshepherd commented:
Merging now, we had support from others in dev meeting to get this in, we're both +1, it tests well, and it fixes something that is currently broken in released DSpace versions.
